### PR TITLE
Implement `-follow`

### LINKED ISSUE #308: Implement `-follow`
Even if this is deprecated, for compat, we should implement it:
```

       -follow
              Deprecated; use the  -L  option  instead.   Dereference  symbolic  links.   Implies
              -noleaf.   The -follow option affects only those tests which appear after it on the
              command line.  Unless the -H or -L option has been specified, the position  of  the
              -follow  option  changes the behaviour of the -newer predicate; any files listed as
              the argument of -newer will be dereferenced if they are symbolic links.   The  same
              consideration applies to -newerXY, -anewer and -cnewer.  Similarly, the -type pred-
              icate will always match against the type of the file that a symbolic link points to
              rather  than  the  link itself.  Using -follow causes the -lname and -ilname predi-
              cates always to return false.
```

### PR DESCRIPTION
**Pull request by hanbings**
Closed #308

May conflict with https://github.com/uutils/findutils/pull/416, better merge #416 first. :)

There are changes as described in `man find`:
(The following changes are: When matching files, match the file pointed to by the file link.)

1. It will use noleaf by default. (But -noleaf has no behavior change yet)
2. changes the behaviour of the -newer predicate.
3. consideration applies to -newerXY, -anewer and -cnewer
4. -type predicate will always match against the type of the file that a symbolic link points to rather than the link itself.
5. causes the -lname and -ilname predicates always to return false.

### CODE REVIEW

**Comment by hanbings** [comment]
Sorry, the test coverage is still not up to target. 
They almost always appear in error handling logic and, therefore, are difficult to cover by test code.
I would like to focus on improving coverage after completing all GNU feature implementations.

File: src/find/matchers/lname.rs [hunk 1]
```diff
@@ -37,17 +37,23 @@ fn read_link_target(file_info: &DirEntry) -> Option<PathBuf> {
...
-        Self { pattern }
+        Self { pattern, follow }
     }
 }
 
 impl Matcher for LinkNameMatcher {
     fn matches(&self, file_info: &DirEntry, _: &mut MatcherIO) -> bool {
+        // -follow: causes the -lname and -ilname predicates always to return false.
+        if self.follow {
+            return false;
```
Comment (tavianator) [inline src/find/matchers/lname.rs:54]: Not really:

```console
tavianator@graphene $ ln -s nowhere broken
tavianator@graphene $ find -lname nowhere
./broken
```

Comment (hanbings) [reply src/find/matchers/lname.rs:54]: Sorry, could you please give me some more detailed steps?
I noticed that the `-follow` parameter is missing, is that `find -follow -lname nowhere`?

Comment (tavianator) [reply src/find/matchers/lname.rs:54]: Whoops I copy pasted the wrong part of my session, but yeah:

```console
$ ln -s nowhere broken
$ find -follow -lname nowhere
./broken
```

Comment (hanbings) [reply src/find/matchers/lname.rs]: Thanks. I have fixed it to the actual behavior of the GNU version in https://github.com/uutils/findutils/pull/420/commits/a0d4e468676299fde7d8ae63108864a71f9770b4

**Review by tavianator — CHANGES_REQUESTED** [review]

File: src/find/matchers/time.rs [hunk 2]
```diff
@@ -29,23 +29,54 @@ fn get_time(matcher_io: &mut MatcherIO, today_start: bool) -> SystemTime {
...
 /// This matcher checks whether a file is newer than the file the matcher is initialized with.
 pub struct NewerMatcher {
     given_modification_time: SystemTime,
+    follow: bool,
 }
 
 impl NewerMatcher {
-    pub fn new(path_to_file: &str) -> Result<Self, Box<dyn Error>> {
+    pub fn new(path_to_file: &str, follow: bool) -> Result<Self, Box<dyn Error>> {
         let metadata = fs::metadata(path_to_file)?;
```
[the remark this answers is not in the document]
Comment (hanbings) [reply src/find/matchers/time.rs]: Thanks, I opened an issue tracking this: https://github.com/uutils/findutils/issues/431, and I'd like to open some new PRs to address it. :)

File: src/find/matchers/time.rs [hunk 3]
```diff
@@ -29,23 +29,54 @@ fn get_time(matcher_io: &mut MatcherIO, today_start: bool) -> SystemTime {
...
+            follow,
         })
     }
 
     /// Implementation of matches that returns a result, allowing use to use try!
     /// to deal with the errors.
     fn matches_impl(&self, file_info: &DirEntry) -> Result<bool, Box<dyn Error>> {
-        let this_time = file_info.metadata()?.modified()?;
+        let path = file_info.path();
+        let this_time = if self.follow && path.is_symlink() {
```
[the remark this answers is not in the document]
Comment (hanbings) [reply src/find/matchers/time.rs:78]: The corresponding logic for using `symlink_metadata` has been added in the current commit, and I have also opened an issue to track the problem of `metadata` caching: https://github.com/uutils/findutils/issues/430. Thanks!

File: src/find/matchers/lname.rs [hunk 4]
```diff
@@ -27,48 +33,43 @@ fn read_link_target(file_info: &DirEntry) -> Option<PathBuf> {
                 )
                 .unwrap();
             }
-
```
Comment (tavianator) [inline src/find/matchers/lname.rs:30]: Should probably keep all these deleted blank lines

**Review by tavianator — CHANGES_REQUESTED** [review]
Presumably you want to call [`WalkDir::follow_links(config.follow)`](https://docs.rs/walkdir/latest/walkdir/struct.WalkDir.html#method.follow_links) at some point?

Also, you could implement `-L` at the same time.

**Review by tavianator — APPROVED** [review]

**Comment by tavianator** [comment]
@sylvestre #436 contains a squashed version of this PR if you'd rather avoid fixing the tests here

**Comment by tavianator** [comment]
This can be closed now that #436 was merged

**Pull request closed without merging**